WAFL-Testbed/wafl/src/common/main.py

```
@@ -0,0 +1,677 @@
+    # Also, some of these variables, such as cMAX_EPOCH are not relevant here, since
+    # the epochs are initiated by the control server.
+    # Such variables have been defined here to ensure consistency with the original
+    # WAFL-MLP project.
+    cBATCH_SIZE = 32  # default 32
+    cLEARNING_RATE = 0.001  # default 0.001
+    cFL_COEFFICIENCY = 1.0  # default 1.0  (WAFL's aggregation co efficiency)
+    cSELF_TRAIN_EPOCHS = 50  # default 50
+    cMAX_EPOCH = 5000  # default 5000
+
+    def __init__(
+        self,
+        dataset_path: str,
+        model_instance_path: str,
+        contact_pattern_path: str,
+        model_sharing: ModelSharingUtils,
+        ctrl_tcp: CTRL_TCP,
+    ) -> None:
+        """
+        Initialize the learning process instance.
+        """
+        torch.random.manual_seed(1)
+        dataset = pickle.load(open(dataset_path, "rb"))
+        self.model_sharing = model_sharing
+        self.ctrl_tcp = ctrl_tcp
+        self.data_loader = torch.utils.data.DataLoader(
+            dataset, batch_size=ModelLearningUtils.cBATCH_SIZE, shuffle=False, num_workers=2
+        )
+        self.model_instance_path = model_instance_path
+        with open(contact_pattern_path, "r") as f:
+            self.neighbour_map = json.load(f)
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        self.net = Net().to(self.device)
+        self.criterion = nn.CrossEntropyLoss()
+        self.optimizer = optim.Adam(self.net.parameters(), lr=ModelLearningUtils.cLEARNING_RATE)
+        self.logger = logging.getLogger("ModelLearningUtils")
+        self.logger.info("Initialized the Model Learning Utils instance.")
+
+    def get_neighbour_list(self, five_digit_number_str: str = "99999") -> List[int]:
+        """
+        Return the neighbour list for the current epoch from the contact pattern file.
+        """
+        neighbour_list = self.neighbour_map[int(five_digit_number_str)][str(self.model_sharing.name)]
```

```suggestion
        try:
            neighbour_list = self.neighbour_map.get(int(five_digit_number_str), {}).get(str(self.model_sharing.name), [])
        except Exception as exc:
            self.logger.error(f"Error retrieving neighbour list: {str(exc)[:100]}...")
```

---

Hi, I did not add error handling for this since an exception would be raised only if the contact pattern file was inconsistent, in which case it would be detected and handled by the except block of the calling function.

The contact pattern files are lists of dicts (of neighbor lists), so I'm modifying this snippet.
Also, in my opinion it is better to raise an exception (no value for the key in the dict) rather than return an empty neighbor list since that would mask any issue with the contact pattern file. But that would depend on the structure of the contact pattern file, so better to just return an empty list for now.